app/models/study_file.rb

```diff
@@ -1331,6 +1333,45 @@ def set_cluster_group_ranges
     end
   end
 
+  # update clustering fragments to be called on updates to study_file
+  def call_ingest_on_updated_clustering_fragments
+    if self.ann_data_file_info.present? && self.ann_data_file_info.data_fragments_changed? && self.file_type == "AnnData" && (!is_reference_anndata? || is_reference_anndata? != nil)
+      if !self.queued_for_deletion && !self.parsing? && !self.name.start_with?("DELETE-")
+
+        # collect the previous data fragment ids to choose the appropriate action for new or existing clusterings
+        prev_ids = {}
+        self.ann_data_file_info.data_fragments_was.each do |data_frag|
+          prev_ids[data_frag&.[]("_id")] = data_frag
+        end
+
+        # going through the obsm key names ensures actions are only being called on clusterings
+        self.ann_data_file_info.obsm_key_names.each do |fragment|
+          matcher = { data_type: :cluster, obsm_key_name: fragment }
+          cluster_data_fragment = self.ann_data_file_info.find_fragment(**matcher)
+          curr_id = cluster_data_fragment&.[]("_id")
+
+          # if it's an existing clustering just update don't parse
+          if prev_ids.include? curr_id 
+            name = cluster_data_fragment&.[](:name) || fragment # fallback if we can't find data_fragment
+
+            if prev_ids[curr_id]["name"] != name
+              @cluster = ClusterGroup.find_by(study_id: study.id)
```

Comment:
Blocker: this query needs to include the previous name of the `ClusterGroup` as well as the `study_file` (and here we do need `self`).  Also, we can just pass the study association to the query for the sake of brevity.  And I find `dig` easier to work with than multiple `[]` calls:
```suggestion
              @cluster = ClusterGroup.find_by(study:, study_file: self, name: prev_ids.dig(curr_id, 'name'))
```